### PR TITLE
New Scalable Side Nav design Implementation

### PR DESCRIPTION
**Pull request by hiba-aldalaty**
## Hey, I just made a Pull Request!

https://github.com/backstage/backstage/issues/6933
This PR implements the new scalable side-nav design introduced in the above issue. 
The changes include:
- Optional: you can add the new expand-button to expand the sidebar when it is clicked, and you can disable slide-out on hover. This can be less distracting.
- You can add items into the sidebar which have their own submenus.
- Submenus can hold submenu items like further links to navigate around backstage or even multi-category items
- Highlight on hover - items in the sidebar and in submenus are now highlighted when hovered on.

Design:  [SidenavFigma](https://www.figma.com/proto/df353LDD1WiXh78M9a3jK9/Side-Navigation?page-id=229%3A567&node-id=229%3A3625&viewport=267%2C48%2C1&scaling=contain&starting-point-node-id=229%3A3625&show-proto-sidebar=3)

Video Walkthrough: [SidenavVideoWalkthrough](https://drive.google.com/file/d/1gciX9w0qw1RARMNNGfwZuEgkWotxrr5P/view?usp=sharing)

Screenshot 2021-11-04 at 16 27 04
Screenshot 2021-11-04 at 16 26 49
Screenshot 2021-11-04 at 16 26 39
Screenshot 2021-11-04 at 16 24 19
Screenshot 2021-11-04 at 16 26 12
hubusercontent.com/72443252/140379382-5b003cc0-0be6-4ba6-86de-68b9f211e23b.png">
Screenshot 2021-11-04 at 16 28 52

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
Just noting here that we discussed the question on discord, so that future goalies know. The intent is that user and adopter impact is none at all, as the relevant options are added.

**Comment by hiba-aldalaty** [comment]
Hey @freben , I have made a couple of changes to make sure the change does not cause any impact when migrating. This PR should be ready now.

**Comment by guillermomanzo** [comment]
hey @benjdlambert @freben We are hoping to get this integrated by EOW. any chance of that happening. hoping is the key word ;)

**Review by benjdlambert — COMMENTED** [review]
Screenshot 2021-11-17 at 15 52 18

One other small note, is that the hit slop on the submenu items is pretty small. Can we make this full width and full height.

**Comment by benjdlambert** [comment]
@guillermomanzo @hiba-aldalaty sorry it's taken a while to get a review of this done, I've unfortunately been sick for the past week or so, and still am to a certain degree. We usually release in the afternoon, so if you can get this all fixed up tomorrow for early afternoon there's a possibility that it will get shipped tomorrow, but otherwise it's going to fall into the release next week. We've also got some other work going on in parallel that we're holding off merging #6905 until this falls in too as it's easier for us to fix the conflicts on our end rather than you having to fix that up too. Reach out if you have any further questions 👍

**Comment by hiba-aldalaty** [comment]
Appreciate your feedback @benjdlambert . It might have to fall to next weeks release but I'll try tackle it as soon as I can. Get well soon!

**Review by tudi2d — COMMENTED** [review]
Great feature - really looking forward to seeing this & it's also going to fit well with the mobile sideba #6905. Just a few smaller things from my side.

File: packages/core-components/api-report.md
Comment (hiba-aldalaty) [reply packages/core-components/api-report.md:2101]: That is a good point. I did begin considering this while implementing but as this is the first iteration, I opted for the simplest route to get started and making it more flexible in the next iterations. 
The current way allows us to make sure the sidebar item remains active/highlighted when a dropdown item is selected, using the 'to' value. If dropdownItems was JSX, we'd have to consider how we're going to ensure this.

Comment (hiba-aldalaty) [reply packages/core-components/api-report.md:2101]: We can allow rendering other components by changing `title` in `DropDownItem` type to be JSX and the `to` property could be optional. Would allow the same functionality, just doesn't have to be text.

File: packages/core-components/src/layout/Sidebar/Items.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Items.tsx:195]: Thanks Philipp, left a comment about this above

File: packages/core-components/src/layout/Sidebar/Bar.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:95]: isPinned=false still allows expanding on hover
I think disableExpandOnHover provides slightly different functionality - it prevents the sidebar from expanding completely unless it is pinned or an expand button is added to control it via a click instead of mouseover.

**Comment by hiba-aldalaty** [comment]
Hey @tudi2d @benjdlambert , 
Thanks for your comments. I have made a few changes using your feedback, including:
- Tackling some of the warnings in api-report
- Removing unnecessary props or moving them elsewhere
- Decreasing api surface by using setOpen as suggested
- Improving conditional class statements
- Rename SubmenuItem to SidebarSubItem

Part of the changes included a slight refactor to allow the submenu items to be wrapped inside the submenu component directly. I think, in general, this should make it easier to understand, use and read. Should also make it more scalable.
SidebarItem -> Submenu -> Submenu Items.

I would like to note, SidebarItems and ‘SidebarSubItems’ are different in where they are used etc… SidebarSubItems should be used solely in submenus and never in the sidebar, and vice versa for SidebarItems. For this reason I made them as a separate component to SidebarItem.

**Comment by benjdlambert** [comment]
@hiba-aldalaty perfect, thanks for the explanation and the recent changes. We will get round to re-reviewing again soon, in the meantime there's also some failing tests which could do with looking at :)

File: packages/theme/api-report.md
Comment (Rugvip) [inline packages/theme/api-report.md:44]: Unfortunately this is a breaking change that we can't introduce straight away. Do you see some other option here like using something else that already exists in the palette? Perhaps something from the base MUI palette? Otherwise we'd need to make these optional along with a fallback.

**Review by Rugvip — COMMENTED** [review]
Thank you! Another round of feedback with a well rounded collection of naming nits and other things x)

**Comment by guillermomanzo** [comment]
Lots of great feedback and discussion here. Hiba is one of our Expedia Group interns, and is definitely learning a ton from all of y'all. :)

**Comment by sethhoward-dd** [comment]
Hi, Seth from DoorDash here, we're very interested in this PR landing as we were about to custom build the same thing. Let us know if there's anything we can do to help!

File: packages/theme/api-report.md
Comment (hiba-aldalaty) [reply packages/theme/api-report.md:44]: Sadly I can't seem to find another suitable option. Do you have any suggestions on how we would go about making these optional?

File: packages/core-components/src/layout/Sidebar/Items.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Items.tsx:417]: That sounds great, thank you!

File: packages/core-components/src/layout/Sidebar/Sidebar.stories.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Sidebar.stories.tsx]: Unfortunately this was a hard one to try find a solution without reaching some compromise. My latest commit offers a solution but could use another opinion

File: packages/core-components/src/layout/Sidebar/Sidebar.stories.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Sidebar.stories.tsx]: Definitely looks better 👍🏻

File: packages/core-components/src/layout/Sidebar/Bar.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:25]: Some of them I wasn't able to find and some I could not import even though they should be in material-ui. Getting some confirmation on this, will get back to you about it soon

**Comment by hiba-aldalaty** [comment]
@Rugvip Hi Patrik, 
Apologies, took me a little longer to get back with some changes as I was off most of yesterday. Appreciate your feedback! I made some changes and left a couple of comments. Feel free to add any further suggestions.

File: packages/theme/api-report.md
Comment (hiba-aldalaty) [reply packages/theme/api-report.md]: Update: I made them into optional props, should resolve this issue

File: packages/core-components/src/layout/Sidebar/Bar.tsx
Comment (Rugvip) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:25]: Alright, thanks! We'd like to avoid adding any additional items, both because MUI already offers a rich set of icons, but also because we'd need to make sure licensing of any added items is done correctly, which can be a can of worms.

**Review by Rugvip — COMMENTED** [review]
Thanks! 👍, another round of review meanwhile, we're getting there :grin:

File: packages/storybook/.storybook/apis.js
Comment (hiba-aldalaty) [reply packages/storybook/.storybook/apis.js:34]: Its necessary to use the useElementFeature function because it relies on featureFlagsApiRef

File: packages/core-components/api-report.md
Comment (hiba-aldalaty) [reply packages/core-components/api-report.md:1985]: whats 4 extra characters 😅

File: packages/core-components/src/layout/Sidebar/Bar.tsx
Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:25]: > 

image
When I try to import many of the icons that should be in @material-ui/icons, I get this error. Can't seem to figure out why because the documentation for the material-ui package say it should be there.

Comment (freben) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:25]: Those are from the docs for v5 of material-ui - we are sadly still on v4. The imports will be on the form `@material-ui/icons/X`. The docs are here:

https://v4.mui.com/

And the icons are here

https://v4.mui.com/components/material-icons/

Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:25]: Got it, thank you!! Not an issue, I replaced most of the icons with ones in v4 as they're only really used in the storybook. I created a double arrow from scratch with the existing single arrows from mui so it still matches the design 🌟

Comment (hiba-aldalaty) [reply packages/core-components/src/layout/Sidebar/Bar.tsx:25]: @Rugvip I believe the latest commit should resolve this issue now

File: packages/core-components/src/layout/Sidebar/SidebarSubmenuItem.tsx
Comment (Rugvip) [inline packages/core-components/src/layout/Sidebar/SidebarSubmenuItem.tsx:90]:
```suggestion
export type SidebarSubmenuItemDropdownItem = {
```

Or we make it `dropDownItems`, but I think this makes more sense :p

**Review by Rugvip — COMMENTED** [review]
Thank you the updates, just a few steps left and we're good to go :grin:

Got a tiny nit on the naming that I missed, but otherwise I think implementation is good to go 👍 , and then the merge conflicts need to be resolved

Last bit is to create [changesets](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets) to generate changelog entries and make sure this gets released. In this case I think separate changesets for the core-components and the theme package would make sense. The storybook and test changes in the shortcuts plugin don't require changesets

**Comment by benjdlambert** [comment]
@hiba-aldalaty I've updated the PR with the feedback from @Rugvip as we ideally want to try and get this in the release today. Hope you don't mind. :pray:

**Comment by hiba-aldalaty** [comment]
@benjdlambert Apologies, been on support and haven't had the chance to make the changes. Thats totally fine, thanks for making the change

**Review by freben — APPROVED** [review]

**Review by benjdlambert — APPROVED** [review]

File: packages/core-components/src/layout/Sidebar/SidebarSubmenuItem.tsx
Comment (mwz) [inline packages/core-components/src/layout/Sidebar/SidebarSubmenuItem.tsx:38]: Hi, is there any chance this component's styles could be made overridable so that it's possible to create a custom theme `BackstageOverrides` for this component?

**Pull request merged**
